src/DDoSX/WafClient.php

```diff
@@ -26,14 +33,141 @@ public function create(Waf $waf)
+     */
+    public function getRules($domainName, $page = 1, $perPage = 15, $filters = [])
+    {
+        $page = $this->paginatedRequest('v1/domains/' . $domainName . '/waf/rules', $page, $perPage, $filters);
+
+        $page->serializeWith(function ($item) {
+            return $this->serializeRule($item);
+        });
+
+        return $page;
+    }
+
+    /**
+     * Get the rule for a domain by its ID
+     * @param $domainName
+     * @param $ruleId
+     * @return WafRule
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function getRuleById($domainName, $ruleId)
+    {
+        $response = $this->get('v1/domains/' . $domainName . '/waf/rules/' . $ruleId);
+        $body = $this->decodeJson($response->getBody()->getContents());
+
+        return $this->serializeRule($body->data);
+    }
+
+    /**
+     * Get a paginated response of advanced rules for a domain
+     * @param $domainName
+     * @param int $page
+     * @param int $perPage
+     * @param array $filters
+     * @return int|\UKFast\SDK\Page
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function getAdvancedRules($domainName, $page = 1, $perPage = 15, $filters = [])
+    {
+        $page = $this->paginatedRequest(
+            'v1/domains/' . $domainName . '/waf/advanced-rules',
+            $page,
+            $perPage,
+            $filters
+        );
+
+        $page->serializeWith(function ($item) {
+            return $this->serializeAdvancedRule($item);
+        });
+
+        return $page;
+    }
+
+    /**
+     * Get an advanced rule for a domain by its ID
+     * @param $domainName
+     * @param $ruleId
+     * @return WafAdvancedRule
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function getAdvancedRuleById($domainName, $ruleId)
+    {
+        $response = $this->get('v1/domains/' . $domainName . '/waf/advanced-rules/' . $ruleId);
+        $body = $this->decodeJson($response->getBody()->getContents());
+
+        return $this->serializeAdvancedRule($body->data);
+    }
+
+    /**
+     * Get the rulesets for a domain
+     * @param $domainName
+     * @return array
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function getRulesets($domainName)
+    {
+        $response = $this->get('v1/domains/' . $domainName . '/waf/rulesets');
+
+        $body = $this->decodeJson($response->getBody()->getContents());
+
+        return array_map(function ($item) {
+            return $this->serializeRuleset($item);
+        }, $body->data);
+    }
+
+    protected function serializeWaf($raw)
+    {
+        return new Waf($this->apiToFriendly($raw, self::WAF_MAP));
+    }
+
+    protected function serializeRule($raw)
+    {
+        return new WafRule($this->apiToFriendly($raw, self::RULE_MAP));
+    }
+
+    protected function serializeAdvancedRule($raw)
+    {
+        return new WafAdvancedRule($this->apiToFriendly($raw, self::RULE_MAP));
```

Comment:
```suggestion
        return new WafAdvancedRule($this->apiToFriendly($raw, self::ADVANCED_RULE_MAP));
```
probably a copy pasta - we can either keep the rule map here and remove the definition of the advanced rule map, or use the suggestion

Reply:
We'll keep for now as it then keeps it inline with all the other clients, It is more than likely a c+p job, cheers